### PR TITLE
`HDF5` file structure reorganisation, and `InterfaceComponent` readability

### PR DESCRIPTION
**Pull request by samcunliffe**
This PR is a single `git cherry-pick` of the merge commit of #280 but now (eventually) targeting `main`.

Part of 

* #181 

Continues from
1. #304 
2. #310 
3. #311 

Supercedes
* #193
* #250 

## Actual code notes and change description

... are in 
* #280

For future working, I think the "move files" and "add tests" steps here should have been separate PRs.
- [ ] Check the diff with files on `main` to ensure none of the review improvements are lost.

### CODE REVIEW

File: tdms/include/hdf5_io/hdf5_reader.h [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+    std::vector<hsize_t> dimensions = shape_of(dataset_name);
+    if (dimensions.size() != 2) {
+      throw std::runtime_error(
+              "Cannot read " + dataset_name + " into a 2D matrix, it has " +
+              std::to_string(dimensions.size()) + " dimensions");
+    }
+    int n_rows = dimensions[0];
+    int n_cols = dimensions[1];
+
+    SPDLOG_DEBUG("n_rows = {}; n_cols = {}", n_rows, n_cols);
```
Comment (samcunliffe) [inline tdms/include/hdf5_io/hdf5_reader.h:70]:
```suggestion
    spdlog::debug("n_rows = {}; n_cols = {}", n_rows, n_cols);
```

File: tdms/src/hdf5_io/hdf5_base.cpp [hunk 2]
```diff
@@ -1,37 +1,30 @@
...
-  H5::DataSet dataset = file_->createDataSet(dataset_name, datatype, dataspace);
-  dataset.write(data, H5::PredType::NATIVE_DOUBLE);
-  spdlog::trace("Write successful.");
+ijk to_ijk(const std::vector<hsize_t> dimensions) {
+  unsigned int rank = dimensions.size();
+  ijk out;
+  if (rank > 0) out.i = (int) dimensions[0];
+  if (rank > 1) out.j = (int) dimensions[1];
+  if (rank > 2) out.k = (int) dimensions[2];
+  if (rank > 3) spdlog::warn("Rank > 3");
```
Comment (samcunliffe) [inline tdms/src/hdf5_io/hdf5_base.cpp:22]: Should be a `std::logic_error` or `std::runtime_error`? Is there ever a time when we are OK with being given a 4-tensor and treating it as a 3-tensor?

File: tdms/src/hdf5_io/hdf5_base.cpp [hunk 3]
```diff
@@ -1,37 +1,30 @@
...
- * HDF5Writer
- */
-void HDF5Writer::write(const std::string &dataset_name, double *data, int size,
-                       hsize_t *dimensions) {
-  spdlog::debug("Writing {} to file: {}", dataset_name, filename_);
-
-  // declare a dataspace
-  H5::DataSpace dataspace(size, dimensions);
-  H5::DataType datatype(H5::PredType::NATIVE_DOUBLE);
+using namespace std;
```
Comment (samcunliffe) [inline tdms/src/hdf5_io/hdf5_base.cpp:14]: I'd advocate removing this and keeping the namespace.

File: tdms/src/hdf5_io/hdf5_base.cpp (hunk 2 above)
Comment (willGraham01) [reply tdms/src/hdf5_io/hdf5_base.cpp:22]: When the trailing dimensions are flat (length 1).

Although this function isn't intended for such use cases (really it's not intended for anything other than converting 3D HDF5 sizes to our native `ijk` structure that we can use throughout TBH)

**Review by willGraham01 — APPROVED** [review]

File: tdms/src/hdf5_io/hdf5_base.cpp (hunk 2 above)
Comment (samcunliffe) [reply tdms/src/hdf5_io/hdf5_base.cpp:22]: Turns out we never use this (it keeps reappearing). I'm going to remove it here.

File: tdms/tests/unit/hdf5_io_tests/test_hdf5_io.cpp [hunk 4]
```diff
@@ -66,13 +55,12 @@ TEST_CASE("Test file I/O construction/destruction.") {
 
     double data[1];
     HDF5Reader fr(tmp.string() + "/this_file_does_exist_but_is_empty.h5");
-    CHECK_THROWS_AS(fr.read("nonexistantdata", data), H5::FileIException);
+    CHECK_THROWS(fr.read("nonexistantdata", data));
```
Comment (samcunliffe) [inline tdms/tests/unit/hdf5_io_tests/test_hdf5_io.cpp:58]: Revert this.
```suggestion
    CHECK_THROWS_AS(fr.read("nonexistantdata", data), H5::FileIException);
```

**Review by willGraham01 — APPROVED** [review]

**Pull request merged**